dashboard/app/controllers/activities_controller.rb

```diff
@@ -56,6 +56,14 @@ def milestone
       end
     end
 
+    if current_user && @script_level && @script_level.stage.lockable?
+      user_level = UserLevel.find_by(user_id: current_user.id, level_id: @script_level.level.id, script_id: @script_level.script.id)
+      # we have a lockable stage, and user_level is locked. disallow milestone requests
+      if user_level.nil? || user_level.locked?(@script_level.stage)
+        return head 403
+      end
+    end
```

Comment:
I wonder if this should be in `User.track_level_progress_sync`?  As it's written now, anything that stores progress without posting to `/milestone` will bypass the `user_level.locked?` check.